File: Net/TemplateMethod/After/TemplateMethod/Beverage.cs
namespace TemplateMethod
{
    using System;

    public abstract class Beverage
    {
        public void PrepareRecipe()
        {
            this.BoilWater();
            this.Brew();
            this.PourInCup();
            this.AddCondiments();
        }

        protected abstract void Brew();

        protected abstract void AddCondiments();

        private void BoilWater()
        {
            Console.WriteLine(BeverageMessages.BoilWater);
        }

        private void PourInCup()
        {
            Console.WriteLine(BeverageMessages.PourInCup);
        }
    }
}

File: Java/TemplateMethod/After/src/main/templatemethod/Beverage.java
package templatemethod;

public abstract class Beverage {
  
	public void prepareRecipe() {
		boilWater();
		brew();
		pourInCup();
		addCondiments();
	}
 
	protected abstract void brew();
  
	protected abstract void addCondiments();
 
	private void boilWater() {
		System.out.println("Boiling water");
	}
  
	private void pourInCup() {
		System.out.println("Pouring into cup");
	}
}
